presto-thrift/src/main/java/com/facebook/presto/thrift/ThriftConnectorIndex.java

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.thrift;
+
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorIndex;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.RecordCursor;
+import com.facebook.presto.spi.RecordSet;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.thrift.clientproviders.PrestoClientProvider;
+import com.facebook.presto.thrift.interfaces.client.ThriftColumnData;
+import com.facebook.presto.thrift.interfaces.client.ThriftRowsBatch;
+import com.facebook.presto.thrift.interfaces.writers.ColumnWriter;
+import com.facebook.presto.thrift.interfaces.writers.ColumnWriters;
+import com.facebook.presto.thrift.pagesources.ThriftIndexPageSource;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class ThriftConnectorIndex
+        implements ConnectorIndex
+{
+    private final PrestoClientProvider clientProvider;
+    private final ThriftConfig config;
+    private final byte[] indexId;
+    private final List<String> inputColumnNames;
+    private final List<ColumnHandle> outputColumns;
+
+    public ThriftConnectorIndex(
+            PrestoClientProvider clientProvider,
+            ThriftConfig config,
+            ThriftIndexHandle indexHandle,
+            List<ColumnHandle> lookupColumns,
+            List<ColumnHandle> outputColumns)
+    {
+        this.clientProvider = requireNonNull(clientProvider, "clientProvider is null");
+        this.config = requireNonNull(config, "config is null");
+        this.indexId = requireNonNull(indexHandle, "indexHandle is null").getIndexId();
+        this.inputColumnNames = requireNonNull(lookupColumns, "lookupColumns is null").stream()
+                .map(handle -> ((ThriftColumnHandle) handle).getColumnName())
+                .collect(toList());
+        this.outputColumns = requireNonNull(outputColumns, "outputColumns is null");
+    }
+
+    @Override
+    public ConnectorPageSource lookup(RecordSet recordSet)
+    {
+        ThriftRowsBatch keys = convertKeys(recordSet, inputColumnNames);
+        return new ThriftIndexPageSource(clientProvider, indexId, keys, outputColumns, config);
+    }
+
+    private static ThriftRowsBatch convertKeys(RecordSet recordSet, List<String> columnNames)
+    {
+        List<Type> columnTypes = recordSet.getColumnTypes();
+        int nColumns = columnTypes.size();
+        checkArgument(nColumns == columnNames.size(), "size of column types and column names doesn't match");
+        List<ColumnWriter> columnWriters = new ArrayList<>(nColumns);
+        int totalRecords = 0;
+        try (RecordCursor cursor = recordSet.cursor()) {
+            // assumes lists are indexable
```

Comment:
I meant that lists support efficient random access. Anyway, I think the comment is a bit redundant there. Removed it.